chemprop/uncertainty/uncertainty_predictor.py

```diff
@@ -0,0 +1,1014 @@
+        self.dataset_type = dataset_type
+        self.loss_function = loss_function
+        self.uncal_preds = None
+        self.uncal_vars = None
+        self.uncal_confidence = None
+        self.individual_vars = None
+        self.num_models = num_models
+        self.uncertainty_dropout_p = uncertainty_dropout_p
+        self.dropout_sampling_size = dropout_sampling_size
+        self.individual_ensemble_predictions = individual_ensemble_predictions
+        self.spectra_phase_mask = spectra_phase_mask
+        self.train_class_sizes = None
+
+        self.raise_argument_errors()
+        self.test_data_loader = test_data_loader
+        self.calculate_predictions()
+
+    def raise_argument_errors(self):
+        """
+        Raise errors for incompatible dataset types or uncertainty methods, etc.
+        """
+        pass
+
+    def calculate_predictions(self):
+        """
+        Calculate the uncalibrated predictions and store them as attributes
+        """
+        pass
+
+    def get_uncal_preds(self):
+        """Return the predicted values for the test data."""
+        return self.uncal_preds
+
+    def get_uncal_vars(self):
+        """Return the uncalibrated variances for the test data"""
+        return self.uncal_vars
+
+    def get_uncal_confidence(self):
+        """Return the uncalibrated confidences for the test data"""
+        return self.uncal_confidence
+
+    def get_individual_vars(self):
+        """Return the variances predicted by each individual model in an ensemble."""
+        return self.individual_vars
+
+    def get_individual_preds(self):
+        """Return the value predicted by each individual model in an ensemble."""
+        return self.individual_preds
+
+    def get_uncal_output(self):
+        """Return the uncalibrated uncertainty outputs for the test data"""
+        pass
+
+
+class NoUncertaintyPredictor(UncertaintyPredictor):
+    def __init__(
+        self,
+        test_data: MoleculeDataset,
+        test_data_loader: MoleculeDataLoader,
+        models: Iterator[MoleculeModel],
+        scalers: Iterator[StandardScaler],
+        num_models: int,
+        dataset_type: str,
+        loss_function: str,
+        uncertainty_dropout_p: float,
+        dropout_sampling_size: int,
+        individual_ensemble_predictions: bool = False,
+        spectra_phase_mask: List[List[bool]] = None,
+    ):
+        super().__init__(
+            test_data=test_data,
+            test_data_loader=test_data_loader,
+            models=models,
+            scalers=scalers,
+            num_models=num_models,
+            dataset_type=dataset_type,
+            loss_function=loss_function,
+            uncertainty_dropout_p=uncertainty_dropout_p,
+            dropout_sampling_size=dropout_sampling_size,
+            individual_ensemble_predictions=individual_ensemble_predictions,
+            spectra_phase_mask=spectra_phase_mask,
+        )
+        self.label = "no_uncertainty_method"
```

Comment:
move to class attribute or property if it's the same for each class instance

Reply:
same goes for all the classes below

Reply:
then you can remove the `__init__()` method of the subclass entirely

Reply:
Oh that is much better as an abstract property